src/jit/lsra.cpp

```diff
@@ -5667,16 +5670,30 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
 
     if (availablePhysRegInterval != nullptr)
     {
+        RegRecord* physRegToUnassign = availablePhysRegInterval;
+#ifdef _TARGET_ARM_
+        if (isAssigned(availablePhysRegInterval, currentInterval->registerType))
+#else
         if (intervalToUnassign != nullptr)
+#endif
         {
-            RegRecord* physRegToUnassign = availablePhysRegInterval;
 #ifdef _TARGET_ARM_
-            // We should unassign a double register if availablePhysRegInterval is part of the double register
-            if (availablePhysRegInterval->assignedInterval->registerType == TYP_DOUBLE &&
-                !genIsValidDoubleReg(availablePhysRegInterval->regNum))
-                physRegToUnassign = findAnotherHalfRegRec(availablePhysRegInterval);
+            if ((currentInterval->registerType == TYP_DOUBLE) ||
+                (intervalToUnassign != nullptr && intervalToUnassign->registerType == TYP_DOUBLE))
+            {
+                if (!genIsValidDoubleReg(availablePhysRegInterval->regNum))
+                {
+                    physRegToUnassign  = findAnotherHalfRegRec(availablePhysRegInterval);
+                    intervalToUnassign = physRegToUnassign->assignedInterval;
+                }
+                unassignDoublePhysReg(physRegToUnassign);
+            }
+            else if (intervalToUnassign != nullptr)
```

Comment:
I still believe that it is better the encapsulate the above logic in `unassignPhysReg()`. And if you used the structure I suggested in my previous comment, I believe it could subsume the existing `unassignDoublePhysReg()` and be simpler:
```
Interval*  intervalToSpill = regRec->assignedInterval;
RegRecord* secondRegRec = nullptr;

if (newRegType == TYP_DOUBLE)
{
    secondRegRec = findAnotherHalfRegRec(regRec);
}
else if (newRegType == TYP_FLOAT)
{
    if ((intervalToSpill != nullptr) && (intervalToSpill->registerType == TYP_DOUBLE))
    {
        secondRegRec = findAnotherHalfRegRec(regRec);
    }
}
if (intervalToSpill != nullptr)
{
    unassignPhysReg(regRec, (intervalToSpill->recentRefPosition);
}
if ((secondRegRec != nullptr) && (secondRegRec->assignedInterval != nullptr))
{
    unassignPhysReg(secondRegRec, (secondRegRec->assignedInterval->recentRefPosition);
}
```